packages/core/src/utils/security/index.ts

```diff
@@ -1,55 +1,115 @@
   );
 
-  if (firstLeftValue.code !== 401 || invalidSecuritySchemes.length === 1) {
-    return firstLeftValue;
+  const invalidSecurity =
+    firstLeftValue.code !== 401
+      ? firstLeftValue
+      : gatherWWWAuthHeader(invalidSecuritySchemes, ['tags'], firstLeftValue);
+
+  return some(invalidSecurity);
+}
+
+function gatherValidationResults(securitySchemes: SecurityScheme[][], someInput: unknown, resource: unknown) {
+  const authResults = getAuthResults(securitySchemes, someInput, resource);
+
+  const validSecurityScheme = authResults.find((authRes: Array<Either<IPrismDiagnostic, unknown>>) =>
+    authRes.every(isRight),
+  );
+  const invalidSecuritySchemes = authResults.filter((authRes: Array<Either<IPrismDiagnostic, unknown>>) =>
+    authRes.some(isLeft),
+  );
+
+  const firstLeft = invalidSecuritySchemes[0] && invalidSecuritySchemes[0].find(isLeft);
+
+  if (!validSecurityScheme && firstLeft) {
+    return gatherInvalidResults(firstLeft, invalidSecuritySchemes);
   } else {
-    const allWWWAuthHeaders = invalidSecuritySchemes.reduce((accumulator, currentValue) => {
+    return none;
+  }
+}
+
+function gatherWWWAuthHeader(
+  authResults: Array<Array<Either<IPrismDiagnostic, unknown>>>,
+  pathToHeader: string[],
+  firstAuthErr: IPrismDiagnostic,
+) {
+  const flattenedAuthResults = flatten(authResults);
+
+  if (flattenedAuthResults.length === 1) {
+    return firstAuthErr;
+  } else {
+    const wwwAuthenticateHeaders = flattenedAuthResults.map(authResult => {
       return pipe(
-        currentValue,
-        mapLeft(err => [accumulator, get(err, ['headers', 'WWW-Authenticate'])].filter(identity).join(', ')),
-        fold(authHeader => authHeader || '', () => ''),
+        authResult,
+        fold(result => result.tags || [], noop),
       );
-    }, '');
+    });
+
+    const firstAuthErrWithAuthHeader = set(pathToHeader, flatten(wwwAuthenticateHeaders), firstAuthErr);
+
+    return wwwAuthenticateHeaders.every(identity) ? firstAuthErrWithAuthHeader : firstAuthErr;
+  }
+}
+
+function getAuthResult(
+  firstAuthErrAsLeft: Left<IPrismDiagnostic>,
+  authResult: Array<Either<IPrismDiagnostic, unknown>>,
+) {
+  const firstAuthErr: IPrismDiagnostic = pipe(
+    firstAuthErrAsLeft,
+    fold<IPrismDiagnostic, IPrismDiagnostic, IPrismDiagnostic>(identity, identity),
+  );
+
+  const invalidResultWithAuthHeader = gatherWWWAuthHeader([authResult], ['tags'], firstAuthErr);
+
+  return [left(invalidResultWithAuthHeader)];
+}
+
+function getAuthResults(securitySchemes: SecurityScheme[][], someInput: unknown, resource: unknown) {
+  return securitySchemes.map((securitySchemePairs: SecurityScheme[]) => {
+    const authResult = securitySchemePairs.map((securityScheme: SecurityScheme) => {
+      const schemeHandler = securitySchemeHandlers.find(handler => handler.test(securityScheme));
+
+      return schemeHandler
+        ? schemeHandler.handle(someInput, securityScheme.name, resource)
+        : left({
```

Comment:
@XVincentX, nah, it should be a left with an object structurally equivalent to IPrismDiagnostic. Did this because we were returning an option of IPrismDiagnostic or Error, so yeah, that `or` was  annoying